torch/csrc/Exceptions.cpp

```diff
@@ -137,5 +137,73 @@ ValueError::ValueError(const char *format, ...) {
   va_end(fmt_args);
 }
 
+void PyWarningHandler::py_warning_handler(
+    const c10::SourceLocation& source_location,
+    const std::string& msg) {
+  warning_buffer.push({source_location, msg});
+}
+
+PyWarningHandler::warning_buffer_t PyWarningHandler::warning_buffer = PyWarningHandler::warning_buffer_t();
+
+EnforceWarningBuffer::EnforceWarningBuffer() noexcept(true): prev_handler(c10::Warning::get_warning_handler()) {
+  c10::Warning::set_warning_handler(&PyWarningHandler::py_warning_handler);
+}
+
+/// See NOTE [ Conversion Cpp Python Warning ] for noexcept justification
+EnforceWarningBuffer::~EnforceWarningBuffer() noexcept(false) {
+  c10::Warning::set_warning_handler(prev_handler);
+
+  auto& warning_buffer = PyWarningHandler::warning_buffer;
+
+  if(warning_buffer.size() > 0) {
+    AutoGIL gil;
+
+    PyObject *ptype, *pvalue, *ptraceback;
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+
+    if(ptype) {
+      // A python error happened after the warning
+      // Simply handle with the cpp handler
+      while(warning_buffer.size() > 0) {
+        auto warning = warning_buffer.front();
```

Comment:
`auto warning = std::move(warning_buffer.front());` here is better, it avoids an unnecessary copy assignment

Reply:
But tbh, C++ `queue` is really bad and there isn't really a good reason to use it here: you can just stuff everything on a vector, then process the vector front to back and reset() it to empty when you're all done. (You don't need the ability to incrementally remove items; you only need to remove them all at once in the end.)